File: phbir/ph_localization/report/bir_2307/bir_2307.py
# Copyright (c) 2013, SERVIO Technologies and contributors
# For license information, please see license.txt

import frappe
from frappe.utils import (getdate)
from frappe import _

def execute(filters=None):
    columns, data = [], []
    data = get_data(filters.company, filters.supplier, filters.doctype, filters.purchase_invoice, filters.payment_entry, filters.from_date, filters.to_date)
    columns = get_columns()
    return columns, data

# works on net total only
# TODO: custom tax base
def get_data(company, supplier, doctype, purchase_invoice, payment_entry, from_date, to_date):
    result = []
    data_pi = []
    data_pe = []
    pi_condition = ''
    pe_condition = ''
    
    if doctype == "Purchase Invoice" and purchase_invoice:
        pi_condition += ' AND pi.name = %s' % frappe.db.escape(purchase_invoice)

    if doctype == "Payment Entry" and payment_entry:
        pe_condition += ' AND pe.name = %s' % frappe.db.escape(payment_entry)


    if not payment_entry:
        data_pi = frappe.db.sql("""
            SELECT
                temp.payment_type,
                temp.atc,
                temp.description,
                SUM(temp.month_1) AS month_1,
                SUM(temp.month_2) AS month_2,
                SUM(temp.month_3) AS month_3,
                SUM(temp.total) AS total,
                SUM(temp.tax_withheld_for_the_quarter) AS tax_withheld_for_the_quarter
            FROM
            (
                SELECT 
                    (CASE 
                        WHEN ptac.atc IN (SELECT atc FROM `tabATC` WHERE LEFT(atc, 2) IN ('WI', 'WC')) 
                            THEN 'Income Payment' 
                        WHEN ptac.atc IN (SELECT atc FROM `tabATC` WHERE LEFT(atc, 2) IN ('WB', 'WV'))  
                            THEN 'Money Payment' 
                    END) AS payment_type,
                    ptac.atc AS atc,
                    atc.description AS description,
                    pi.name AS document_no,
                    (CASE WHEN MONTH(pi.posting_date) IN (1, 4, 7, 10) THEN pi.base_net_total ELSE 0 END) AS month_1,
                    (CASE WHEN MONTH(pi.posting_date) IN (2, 5, 8, 11) THEN pi.base_net_total ELSE 0 END) AS month_2,
                    (CASE WHEN MONTH(pi.posting_date) IN (3, 6, 9, 12) THEN pi.base_net_total ELSE 0 END) AS month_3,
                    pi.base_net_total AS total,
                    ABS(ptac.base_tax_amount) AS tax_withheld_for_the_quarter
                FROM 
                    `tabPurchase Invoice` pi
                LEFT JOIN
                    `tabPurchase Taxes and Charges` ptac
                ON
                    pi.name = ptac.parent
                LEFT JOIN 
                    `tabSupplier` as s
                ON
                    pi.supplier = s.name
                INNER JOIN
                    `tabAccount` a
                ON
                    ptac.account_head = a.name
                INNER JOIN
                    `tabATC` atc
                ON
                    ptac.atc = atc.name
                WHERE
                    pi.docstatus = 1
                    and pi.is_return = 0
                    and ((ptac.base_tax_amount < 0 and ptac.add_deduct_tax != 'Deduct') or (ptac.base_tax_amount >= 0 and ptac.add_deduct_tax = 'Deduct'))
                    and ptac.atc IN (SELECT atc FROM `tabATC` WHERE tax_type_code IN ('WE', 'WB', 'WV') and form_type = '2307')
                    and (a.account_type in ('Tax', 'Payable', '') or a.account_type is NULL)
                    and pi.supplier = %s
                    and pi.posting_date >= %s
                    and pi.posting_date <= %s
                    and pi.company = %s {0}
            ) temp
            GROUP BY
                temp.payment_type,
                temp.atc,
                temp.description
            """.format(pi_condition), (supplier, getdate(from_date), getdate(to_date), company), as_dict=1)
    
        result.extend(data_pi)
    
    if not purchase_invoice:
        data_pe = frappe.db.sql("""
            SELECT
                temp.payment_type,
                temp.atc,
                temp.description,
                SUM(temp.month_1) AS month_1,
                SUM(temp.month_2) AS month_2,
                SUM(temp.month_3) AS month_3,
                SUM(temp.total) AS total,
                SUM(temp.tax_withheld_for_the_quarter) AS tax_withheld_for_the_quarter
            FROM
            (
                SELECT 
                    (CASE 
                        WHEN atac.atc IN (SELECT atc FROM `tabATC` WHERE LEFT(atc, 2) IN ('WI', 'WC')) 
                            THEN 'Income Payment' 
                        WHEN atac.atc IN (SELECT atc FROM `tabATC` WHERE LEFT(atc, 2) IN ('WB', 'WV'))  
                            THEN 'Money Payment' 
                    END) AS payment_type,
                    atac.atc AS atc,
                    atc.description AS description,
                    pe.name AS document_no,
                    (CASE WHEN MONTH(pe.posting_date) IN (1, 4, 7, 10) THEN (pe.base_paid_amount_after_tax - pe.base_total_taxes_and_charges) ELSE 0 END) AS month_1,
                    (CASE WHEN MONTH(pe.posting_date) IN (2, 5, 8, 11) THEN (pe.base_paid_amount_after_tax - pe.base_total_taxes_and_charges) ELSE 0 END) AS month_2,
                    (CASE WHEN MONTH(pe.posting_date) IN (3, 6, 9, 12) THEN (pe.base_paid_amount_after_tax - pe.base_total_taxes_and_charges) ELSE 0 END) AS month_3,
                    (pe.base_paid_amount_after_tax - pe.base_total_taxes_and_charges) AS total,
                    ABS(atac.base_tax_amount) AS tax_withheld_for_the_quarter
                FROM 
                    `tabPayment Entry` pe
                LEFT JOIN
                    `tabAdvance Taxes and Charges` atac
                ON
                    pe.name = atac.parent
                LEFT JOIN 
                    `tabSupplier` as s
                ON
                    pe.party = s.name
                INNER JOIN
                    `tabAccount` a
                ON
                    atac.account_head = a.name
                INNER JOIN
                    `tabATC` atc
                ON
                    atac.atc = atc.name
                WHERE
                    pe.docstatus = 1
                    and pe.payment_type = 'Pay'
                    and pe.party_type = 'Supplier'
                    and ((atac.base_tax_amount < 0 and atac.add_deduct_tax != 'Deduct') or (atac.base_tax_amount >= 0 and atac.add_deduct_tax = 'Deduct'))
                    and atac.atc IN (SELECT atc FROM `tabATC` WHERE tax_type_code IN ('WE', 'WB', 'WV') and form_type = '2307')
                    and (a.account_type in ('Tax', 'Payable', '') or a.account_type is NULL)
                    and pe.party = %s
                    and pe.posting_date >= %s
                    and pe.posting_date <= %s
                    and pe.company = %s {0}
            ) temp
            GROUP BY
                temp.payment_type,
                temp.atc,
                temp.description
            """.format(pe_condition), (supplier, getdate(from_date), getdate(to_date), company), as_dict=1)
    
        result.extend(data_pe)

    return result

def get_columns():
    columns = [
        {
            "fieldname": "payment_type",
            "label": _("Type"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "atc",
            "label": _("ATC"),
            "fieldtype": "Link",
            "options": "ATC",
            "width": 60
        },
        {
            "fieldname": "description",
            "label": _("Description"),
            "fieldtype": "Data",
            "width": 150
        },
        {
            "fieldname": "month_1",
            "label": _("1st Month"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "month_2",
            "label": _("2nd Month"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "month_3",
            "label": _("3rd Month"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "total",
            "label": _("Total"),
            "fieldtype": "Currency",
            "width": 150
        },
        {
            "fieldname": "tax_withheld_for_the_quarter",
            "label": _("Tax Withheld"),
            "fieldtype": "Currency",
            "width": 150
        }
    ]

    return columns

File: phbir/ph_localization/report/bir_2307/bir_2307.js
// Copyright (c) 2016, SERVIO Technologies and contributors
// For license information, please see license.txt
/* eslint-disable */

{% include 'phbir/public/js/utils.js' %}

frappe.query_reports["BIR 2307"] = {
	"filters": [
		{
			fieldname:"company",
			label: __("Company"),
			fieldtype: "Link",
			options: "Company",
			default: frappe.defaults.get_user_default("Company"),
            reqd: 1
		},
        {
            fieldname:"supplier",
            label: __("Supplier"),
            fieldtype: "Link",
            options: "Supplier",
            reqd: 0
        },
        {
            fieldname:"employee",
            label: __("Employee"),
            fieldtype: "Link",
            options: "Employee",
            reqd: 0
        },
        {
            fieldname:"doctype",
            label: __("Document Type"),
            fieldtype: "Select",
            options: ["Purchase Invoice", "Payment Entry", "Expense Claim"],
            reqd: 0, 
			on_change: function() {
                let filter_based_on = frappe.query_report.get_filter_value('doctype');
                frappe.query_report.toggle_filter_display('purchase_invoice', filter_based_on != 'Purchase Invoice');
                frappe.query_report.toggle_filter_display('payment_entry', filter_based_on != 'Payment Entry');
                frappe.query_report.toggle_filter_display('expense_claim', filter_based_on != 'Expense Claim');

                frappe.query_report.toggle_filter_display('employee', filter_based_on != 'Expense Claim');
                frappe.query_report.toggle_filter_display('supplier', filter_based_on == 'Expense Claim');

                if (filter_based_on == 'Purchase Invoice') {
                    frappe.query_report.set_filter_value("payment_entry", "");
                    frappe.query_report.set_filter_value("expense_claim", "");
                    frappe.query_report.set_filter_value("employee", "");
                }

                if (filter_based_on == 'Payment Entry') {
                    frappe.query_report.set_filter_value("purchase_invoice", "");
                    frappe.query_report.set_filter_value("expense_claim", "");
                    frappe.query_report.set_filter_value("employee", "");
                }

                if (filter_based_on == 'Expense Claim') {
                    frappe.query_report.set_filter_value("purchase_invoice", "");
                    frappe.query_report.set_filter_value("payment_entry", "");
                    frappe.query_report.set_filter_value("supplier", "");
                }
			}
        },
        {
            fieldname:"purchase_invoice",
            label: __("Purchase Invoice"),
            fieldtype: "Link",
            default: "",
            options: "Purchase Invoice",
            hidden: 1,
			get_query: () => {
				var supplier = frappe.query_report.get_filter_value('supplier');
				return {
					filters: {
						'supplier': supplier,
						'docstatus': 1
					}
				}
			},
            reqd: 0
        },
        {
            fieldname:"payment_entry",
            label: __("Payment Entry"),
            fieldtype: "Link",
            default: "",
            options: "Payment Entry",
            hidden: 1,
			get_query: () => {
				var supplier = frappe.query_report.get_filter_value('supplier');
				return {
					filters: {
						'party_type': 'Supplier',
						'party': supplier,
						'payment_type': 'Pay',
						'docstatus': 1
					}
				}
			},
            reqd: 0
        },
        {
            fieldname:"expense_claim",
            label: __("Expense Claim"),
            fieldtype: "Link",
            default: "",
            options: "Expense Claim",
            hidden: 1,
			get_query: () => {
				var employee = frappe.query_report.get_filter_value('employee');
				return {
					filters: {
						'employee': employee,
						'docstatus': 1
					}
				}
			},
            reqd: 0
        },
        {
            fieldname:"from_date",
            label: __("From Date"),
            fieldtype: "Date",
            default: moment(frappe.datetime.get_today()).startOf('quarter'),
            reqd: 1

        },
        {
            fieldname:"to_date",
            label: __("To Date"),
            fieldtype: "Date",
            default: moment(frappe.datetime.get_today()).endOf('quarter'),
            reqd: 1
        }
    ],
	onload: function(report) {
		report.page.add_inner_button(__("Print BIR 2307"), function() {
            let filter_values = {
                    'company': frappe.query_report.get_filter_value('company'),
                    'supplier': frappe.query_report.get_filter_value('supplier'),
                    'employee': frappe.query_report.get_filter_value('employee'),
                    'doctype': frappe.query_report.get_filter_value('doctype'),
                    'purchase_invoice': frappe.query_report.get_filter_value('purchase_invoice') ? frappe.query_report.get_filter_value('purchase_invoice') : "",
                    'payment_entry': frappe.query_report.get_filter_value('payment_entry') ? frappe.query_report.get_filter_value('payment_entry') : "",
                    'expense_claim': frappe.query_report.get_filter_value('expense_claim') ? frappe.query_report.get_filter_value('expense_claim') : "",
                    'from_date': frappe.query_report.get_filter_value('from_date'),
                    'to_date': frappe.query_report.get_filter_value('to_date'),
                };
            let u = new URLSearchParams(filter_values).toString();
            
            var bir_form_url = frappe.urllib.get_full_url('/api/method/phbir.ph_localization.bir_forms.bir_2307?' + u + '&response_type=pdf')
            let bir_form = window.open(bir_form_url);
        });

        let filter_based_on = frappe.query_report.get_filter_value('doctype');
        frappe.query_report.toggle_filter_display('purchase_invoice', filter_based_on != 'Purchase Invoice');
        frappe.query_report.toggle_filter_display('payment_entry', filter_based_on != 'Payment Entry');
        frappe.query_report.toggle_filter_display('expense_claim', filter_based_on != 'Expense Claim');
    },
};